src/java.base/share/classes/java/lang/reflect/ProxyGenerator.java

```diff
@@ -952,65 +924,66 @@ public String toString() {
     }
 
     /**
-     * A PrimitiveTypeInfo object contains assorted information about
-     * a primitive type in its public fields.  The struct for a particular
+     * A PrimitiveTypeInfo object contains bytecode-related information about
+     * a primitive type in its instance fields. The struct for a particular
      * primitive type can be obtained using the static "get" method.
      */
-    private static class PrimitiveTypeInfo {
-
-        private static Map<Class<?>, PrimitiveTypeInfo> table = new HashMap<>();
-
-        static {
-            add(byte.class, Byte.class);
-            add(char.class, Character.class);
-            add(double.class, Double.class);
-            add(float.class, Float.class);
-            add(int.class, Integer.class);
-            add(long.class, Long.class);
-            add(short.class, Short.class);
-            add(boolean.class, Boolean.class);
-        }
+    private enum PrimitiveTypeInfo {
+        BYTE(byte.class, 0),
+        CHAR(char.class, 0),
+        DOUBLE(double.class, 3),
+        FLOAT(float.class, 2),
+        INT(int.class, 0),
+        LONG(long.class, 1),
+        SHORT(short.class, 0),
+        BOOLEAN(boolean.class, 0);
 
         /**
          * name of corresponding wrapper class
          */
-        private String wrapperClassName;
+        private final String wrapperClassName;
         /**
          * method descriptor for wrapper class "valueOf" factory method
          */
-        private String wrapperValueOfDesc;
+        private final String wrapperValueOfDesc;
         /**
          * name of wrapper class method for retrieving primitive value
          */
-        private String unwrapMethodName;
+        private final String unwrapMethodName;
         /**
          * descriptor of same method
          */
-        private String unwrapMethodDesc;
+        private final String unwrapMethodDesc;
+        /**
+         * Opcode offset from iload, ireturn, etc.
+         * Order is i, l, f, d, a.
+         */
+        private final int opcodeOffset;
 
-        private PrimitiveTypeInfo(Class<?> primitiveClass, Class<?> wrapperClass) {
+        PrimitiveTypeInfo(Class<?> primitiveClass, int opcodeOffset) {
             assert primitiveClass.isPrimitive();
 
-            /**
-             * "base type" used in various descriptors (see JVMS section 4.3.2)
-             */
-            String baseTypeString =
-                    Array.newInstance(primitiveClass, 0)
-                            .getClass().getName().substring(1);
-            wrapperClassName = dotToSlash(wrapperClass.getName());
+            Wrapper wrapper = Wrapper.forPrimitiveType(primitiveClass);
+            // single-char BaseType descriptor (see JVMS section 4.3.2)
+            String baseTypeString = wrapper.basicTypeString();
+            wrapperClassName = dotToSlash(wrapper.wrapperType().getName());
```

Comment:
```suggestion
            wrapperClassName = wrapper.wrapperType().descriptorString();
```

It may worth to replace similar use of `dotToSlash(c.getName())` pattern.

Reply:
Unfortunately, we want an internal name (`xxx/Abc`) than a field descriptor (`Lxxx/Abc;`). But we can use descriptor string for the valueOf descriptor construction.

Reply:
ah, you're right.